# STEP3: Perform some simple data analysis

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed steps 1 and 2, then skip to the second cell.

In [1]:
# !PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f Data/pagila-schema.sql
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f Data/pagila-data.sql

In [2]:
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila'
DB_USER = 'postgres'
DB_PASSWORD = '123456'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://postgres:123456@127.0.0.1:5432/pagila


### 3NF - Entity Relationship Diagram

<img src="images/pagila-3nf.png"/>

## 3.1 Insight 1:   Top Grossing Movies 
- Payments amounts are in table `payment`
- Movies are in table `film`
- They are not directly linked, `payment` refers to a `rental`, `rental` refers to an `inventory` item and `inventory` item refers to a `film`
- `payment` &rarr; `rental` &rarr; `inventory` &rarr; `film`

### 3.1.1 Films

In [3]:
%%sql
SELECT film_id, title, release_year, rental_rate, rating
FROM film
LIMIT 5;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
5 rows affected.


[(1, 'ACADEMY DINOSAUR', 2006, Decimal('0.99'), 'PG'),
 (2, 'ACE GOLDFINGER', 2006, Decimal('4.99'), 'G'),
 (3, 'ADAPTATION HOLES', 2006, Decimal('2.99'), 'NC-17'),
 (4, 'AFFAIR PREJUDICE', 2006, Decimal('2.99'), 'G'),
 (5, 'AFRICAN EGG', 2006, Decimal('2.99'), 'G')]

### 3.1.2 Payments

In [4]:
%sql SELECT * FROM payment LIMIT 5;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
5 rows affected.


[(16050, 269, 2, 7, Decimal('1.99'), datetime.datetime(2017, 1, 24, 21, 40, 19, 996577, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))),
 (16051, 269, 1, 98, Decimal('0.99'), datetime.datetime(2017, 1, 25, 15, 16, 50, 996577, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))),
 (16052, 269, 2, 678, Decimal('6.99'), datetime.datetime(2017, 1, 28, 21, 44, 14, 996577, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))),
 (16053, 269, 2, 703, Decimal('0.99'), datetime.datetime(2017, 1, 29, 0, 58, 2, 996577, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))),
 (16054, 269, 1, 750, Decimal('4.99'), datetime.datetime(2017, 1, 29, 8, 10, 6, 996577, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200))))]

### 3.1.3 Inventory

In [5]:
%sql SELECT * FROM inventory LIMIT 5;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
5 rows affected.


[(1, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))),
 (2, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))),
 (3, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))),
 (4, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))),
 (5, 1, 2, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200))))]

### 3.1.4 Get the movie of every payment

In [6]:
%%sql
SELECT f.title, p.amount
FROM payment p JOIN rental r ON (p.rental_id = r.rental_id)
JOIN inventory i ON (r.inventory_id = i.inventory_id)
JOIN film f ON (i.film_id = f.film_id)
LIMIT 10

 * postgresql://postgres:***@127.0.0.1:5432/pagila
10 rows affected.


[('SWARM GOLD', Decimal('1.99')),
 ('PACKER MADIGAN', Decimal('0.99')),
 ('SOMETHING DUCK', Decimal('6.99')),
 ('DRACULA CRYSTAL', Decimal('0.99')),
 ('CLOSER BANG', Decimal('4.99')),
 ('DANCES NONE', Decimal('2.99')),
 ('STRAIGHT HOURS', Decimal('1.99')),
 ('MARRIED GO', Decimal('4.99')),
 ('DIVORCE SHINING', Decimal('8.99')),
 ('GOLDMINE TYCOON', Decimal('0.99'))]

### 3.1.5 sum movie rental revenue
TODO: Write a query that displays the amount of revenue from each title. Limit the results to the top 10 grossing titles. Your results should match the table below.

In [7]:
%%sql
SELECT f.title, SUM(p.amount) as revenue
FROM payment p 
JOIN rental r ON (p.rental_id = r.rental_id)
JOIN inventory i ON (r.inventory_id = i.inventory_id)
JOIN film f ON (i.film_id = f.film_id)
GROUP BY f.title
ORDER BY revenue DESC
LIMIT 10

 * postgresql://postgres:***@127.0.0.1:5432/pagila
10 rows affected.


[('TELEGRAPH VOYAGE', Decimal('231.73')),
 ('WIFE TURN', Decimal('223.69')),
 ('ZORRO ARK', Decimal('214.69')),
 ('GOODFELLAS SALUTE', Decimal('209.69')),
 ('SATURDAY LAMBS', Decimal('204.72')),
 ('TITANS JERK', Decimal('201.71')),
 ('TORQUE BOUND', Decimal('198.72')),
 ('HARRY IDAHO', Decimal('195.70')),
 ('INNOCENT USUAL', Decimal('191.74')),
 ('HUSTLER PARTY', Decimal('190.78'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>title</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>TELEGRAPH VOYAGE</td>
        <td>231.73</td>
    </tr>
    <tr>
        <td>WIFE TURN</td>
        <td>223.69</td>
    </tr>
    <tr>
        <td>ZORRO ARK</td>
        <td>214.69</td>
    </tr>
    <tr>
        <td>GOODFELLAS SALUTE</td>
        <td>209.69</td>
    </tr>
    <tr>
        <td>SATURDAY LAMBS</td>
        <td>204.72</td>
    </tr>
    <tr>
        <td>TITANS JERK</td>
        <td>201.71</td>
    </tr>
    <tr>
        <td>TORQUE BOUND</td>
        <td>198.72</td>
    </tr>
    <tr>
        <td>HARRY IDAHO</td>
        <td>195.70</td>
    </tr>
    <tr>
        <td>INNOCENT USUAL</td>
        <td>191.74</td>
    </tr>
    <tr>
        <td>HUSTLER PARTY</td>
        <td>190.78</td>
    </tr>
</tbody></table></div>

## 3.2 Insight 2:   Top grossing cities 
- Payments amounts are in table `payment`
- Cities are in table `cities`
- `payment` &rarr; `customer` &rarr; `address` &rarr; `city`

### 3.2.1 Get the city of each payment

In [8]:
%%sql
SELECT ci.city, p.amount
FROM payment p 
JOIN customer c ON (p.customer_id = c.customer_id)
JOIN address a ON (c.address_id = a.address_id)
JOIN city ci ON (a.city_id = ci.city_id)
LIMIT 10;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
10 rows affected.


[('Sasebo', Decimal('2.99')),
 ('Sasebo', Decimal('0.99')),
 ('Sasebo', Decimal('5.99')),
 ('Sasebo', Decimal('0.99')),
 ('Sasebo', Decimal('9.99')),
 ('Sasebo', Decimal('4.99')),
 ('Sasebo', Decimal('4.99')),
 ('Sasebo', Decimal('0.99')),
 ('Sasebo', Decimal('3.99')),
 ('Sasebo', Decimal('4.99'))]

### 3.2.2 Top grossing cities
TODO: Write a query that returns the total amount of revenue by city as measured by the `amount` variable in the `payment` table. Limit the results to the top 10 cities. Your result should match the table below.

In [9]:
%%sql
SELECT ci.city, SUM(p.amount) as revenue
FROM payment p 
JOIN customer c ON (p.customer_id = c.customer_id)
JOIN address a ON (c.address_id = a.address_id)
JOIN city ci ON (a.city_id = ci.city_id)
GROUP BY ci.city
ORDER BY revenue DESC
LIMIT 10;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
10 rows affected.


[('Cape Coral', Decimal('221.55')),
 ('Saint-Denis', Decimal('216.54')),
 ('Aurora', Decimal('198.50')),
 ('Molodetno', Decimal('195.58')),
 ('Santa Brbara dOeste', Decimal('194.61')),
 ('Apeldoorn', Decimal('194.61')),
 ('Qomsheh', Decimal('186.62')),
 ('London', Decimal('180.52')),
 ('Ourense (Orense)', Decimal('177.60')),
 ('Bijapur', Decimal('175.61'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>Cape Coral</td>
        <td>221.55</td>
    </tr>
    <tr>
        <td>Saint-Denis</td>
        <td>216.54</td>
    </tr>
    <tr>
        <td>Aurora</td>
        <td>198.50</td>
    </tr>
    <tr>
        <td>Molodetno</td>
        <td>195.58</td>
    </tr>
    <tr>
        <td>Apeldoorn</td>
        <td>194.61</td>
    </tr>
    <tr>
        <td>Santa Brbara dOeste</td>
        <td>194.61</td>
    </tr>
    <tr>
        <td>Qomsheh</td>
        <td>186.62</td>
    </tr>
    <tr>
        <td>London</td>
        <td>180.52</td>
    </tr>
    <tr>
        <td>Ourense (Orense)</td>
        <td>177.60</td>
    </tr>
    <tr>
        <td>Bijapur</td>
        <td>175.61</td>
    </tr>
</tbody></table></div>

## 3.3 Insight 3 : Revenue of a movie by customer city and by month 

### 3.3.1 Total revenue by month

In [10]:
%%sql
SELECT EXTRACT(month FROM payment_date) AS month, SUM(amount) as revenue
FROM payment
GROUP BY month
ORDER BY revenue DESC
LIMIT 10;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
5 rows affected.


[(Decimal('4'), Decimal('28559.46')),
 (Decimal('3'), Decimal('23886.56')),
 (Decimal('2'), Decimal('9631.88')),
 (Decimal('1'), Decimal('4824.43')),
 (Decimal('5'), Decimal('514.18'))]

### 3.3.2 Each movie by customer city and by month (data cube)

In [11]:
%%sql
SELECT f.title, ci.city, EXTRACT(month FROM p.payment_date) AS month, p.amount
FROM payment p
JOIN rental r ON (p.rental_id = r.rental_id)
JOIN inventory i ON (r.inventory_id = i.inventory_id)
JOIN film f ON (i.film_id = f.film_id)
JOIN customer c ON (p.customer_id = c.customer_id)
JOIN address a ON (c.address_id = a.address_id)
JOIN city ci ON (a.city_id = ci.city_id)
LIMIT 10;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
10 rows affected.


[('PATIENT SISTER', 'Sasebo', Decimal('1'), Decimal('2.99')),
 ('TALENTED HOMICIDE', 'Sasebo', Decimal('1'), Decimal('0.99')),
 ('MUSKETEERS WAIT', 'Sasebo', Decimal('2'), Decimal('5.99')),
 ('DETECTIVE VISION', 'Sasebo', Decimal('2'), Decimal('0.99')),
 ('FERRIS MOTHER', 'Sasebo', Decimal('2'), Decimal('9.99')),
 ('CLOSER BANG', 'Sasebo', Decimal('2'), Decimal('4.99')),
 ('ATTACKS HATE', 'Sasebo', Decimal('2'), Decimal('4.99')),
 ('SAVANNAH TOWN', 'Sasebo', Decimal('2'), Decimal('0.99')),
 ('YOUTH KICK', 'Sasebo', Decimal('2'), Decimal('3.99')),
 ('ADAPTATION HOLES', 'Sasebo', Decimal('3'), Decimal('4.99'))]

### 3.3.3 Sum of revenue of each movie by customer city and by month

TODO: Write a query that returns the total amount of revenue for each movie by customer city and by month. Limit the results to the top 10 movies. Your result should match the table below.

In [12]:
%%sql
SELECT f.title, ci.city, EXTRACT(month FROM p.payment_date) AS month, SUM(p.amount) AS revenue
FROM payment p
JOIN rental r ON (p.rental_id = r.rental_id)
JOIN inventory i ON (r.inventory_id = i.inventory_id)
JOIN film f ON (i.film_id = f.film_id)
JOIN customer c ON (p.customer_id = c.customer_id)
JOIN address a ON (c.address_id = a.address_id)
JOIN city ci ON (a.city_id = ci.city_id)
GROUP BY title, city, month
ORDER BY month, revenue DESC
LIMIT 10;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
10 rows affected.


[('SHOW LORD', 'Mannheim', Decimal('1'), Decimal('11.99')),
 ('KISSING DOLLS', 'Toulon', Decimal('1'), Decimal('10.99')),
 ('TELEGRAPH VOYAGE', 'Naala-Porto', Decimal('1'), Decimal('10.99')),
 ('AMERICAN CIRCUS', 'Callao', Decimal('1'), Decimal('10.99')),
 ('CASUALTIES ENCINO', 'Warren', Decimal('1'), Decimal('10.99')),
 ('AUTUMN CROW', 'Ashgabat', Decimal('1'), Decimal('9.99')),
 ('MIDSUMMER GROUNDHOG', 'Vaduz', Decimal('1'), Decimal('9.99')),
 ('DAY UNFAITHFUL', 'Baybay', Decimal('1'), Decimal('9.99')),
 ('CALIFORNIA BIRDS', 'Plock', Decimal('1'), Decimal('9.99')),
 ('DARKO DORADO', 'Bhilwara', Decimal('1'), Decimal('9.99'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>title</th>
        <th>city</th>
        <th>month</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>SHOW LORD</td>
        <td>Mannheim</td>
        <td>1.0</td>
        <td>11.99</td>
    </tr>
    <tr>
        <td>AMERICAN CIRCUS</td>
        <td>Callao</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>CASUALTIES ENCINO</td>
        <td>Warren</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>TELEGRAPH VOYAGE</td>
        <td>Naala-Porto</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>KISSING DOLLS</td>
        <td>Toulon</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>MILLION ACE</td>
        <td>Bergamo</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>TITANS JERK</td>
        <td>Kimberley</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>DARKO DORADO</td>
        <td>Bhilwara</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>SUNRISE LEAGUE</td>
        <td>Nagareyama</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>MILLION ACE</td>
        <td>Gaziantep</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
</tbody></table></div>